pages/heracles_explained.md

```diff
@@ -3,101 +3,203 @@ layout: page
+//main.ts
+const main = async () => {  
+const resource = await hydraClient.getResource("http://myapi/");  
+// Do something with resource  
+}  
+main();
+```
+
+The getResource method returns a HypermediaContainer. To keep things simple let’s use a hydra powered API provided by the server of this ecosystem. Installing and running hydrus is pretty straightforward. Follow the Instructions [here](https://github.com/HTTP-APIs/hydrus).
+
+Once we have the server up and running we can use its api. So, let’s use it. For now, let’s just console the resource.
+
+```typescript
+//main.ts
+const main = async () => {  
+const resource = await hydraClient.getResource("http://localhost:8000/api/vocab");  
+console.log('resource', resource);  
+}  
+main();
+```
+
+Now, let’s tell the typescript compiler to compile our code into ES2018. For that run in the terminal:
+
+```bash
+tsc --init --target es2018
+```
+
+Since we have used npm packages, to run them in the browser, we need to bundle them. Let’s do just that. In the terminal run:
+
+```bash
+browserify main.ts -p [ tsify --noImplicitAny ] > bundle.js
+```
+To run bundle.js in browser. Let's quickly create a html file index.html and include that script.
+```html
+<!--index.html -->
+ <!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="UTF-8">
+    <title>title</title>
+  </head>
+  <body>
+  <h1>Heracles.ts Demo </h1>
+  </body>
+  <script src="bundle.js"> </script>
+</html>
+```
+Open the file in browser and in the console you can see the response. It should look something like this.
+
+```json
+{
+"@context": { 
+	"ApiDocumentation": "hydra:ApiDocumentation", 
+	"description": "hydra:description", 
+	"expectsHeader": "hydra:expectsHeader"
+	...
+}​
+"@id": "http://localhost:8080/api/vocab"
+"@type": "ApiDocumentation"
+"description": "API Documentation for the server side system"
+​}
+```
+
+
+### Running in node.js environment
+
+To run the client in the node.js environment, we will not be needing tsify and browserify. Instead let’s just install ts-node to run and compile ts in node.js environment and also @types/node for type definitions for node.js.
+
+```bash
+npm install ts-node -g
+```
+
+```bash
+npm install @types/node
+```
+
+Heracles uses isomorphic fetch to make remote calls. So a little tweak in their code can make heracles usable in node.js environment. So, in the node_modules/@hydra-cg/heracles.ts/src/HydraClientFactory.js
+
+Modify the withDefaults function(on line 40):
+
+```javascript
+//HydraClientFactory.js
+withDefaults() {
+	return  this.withJsonLd().with(new BodyResourceBoundIriTemplateExpansionStrategy_1.default())
+	.withStrictLinks()
+	.with(fetch.bind(window));
+}
+```
+By adding ‘global’ keyword as window may be unavailable in node.js environment.
+
+```typescript
+//HydraClientFactory.js
+withDefaults() {
+return  this.withJsonLd().with(new BodyResourceBoundIriTemplateExpansionStrategy_1.default())
+.withStrictLinks()
+.with(fetch.bind(global || window));
+}
+```
```

Comment:
Since Heracles.ts is not our tool, I don't think we need an installation guide or anything. Let me explain why this page is there, as you already know Heracles.ts is the agent built by hydraCG in ts, this page is supposed to be a brief explanation how Heracles.ts is structured, built, what features it has, etc so that developers can take design clues from Heracles.ts architecture in order to improve our own agent. It would be awesome if you could structure this page keeping this goal in mind.

Reply:
Should I remove both browser and node.js installation instructions? The reason why i put it there was because running and seeing things solidified my understanding. So, i thought it can help others too especially people who are just starting out.

Reply:
I guess we can keep it. Add this to the bottom of the page or somewhere nit on the top